### PR TITLE
restrict artifacts greater than size threshold for trial

### PR DESCRIPTION
**Pull request by anandjakhaniya**
**Jira Ticket:** https://vertaai.atlassian.net/browse/VR-7371

### CODE REVIEW

**Review by ravishetye — COMMENTED** [review]
as per a slack discussion we need to pass on ResourceExhausted as the standard error code and the message with a predefined prefix

File: backend/src/main/java/ai/verta/modeldb/artifactStore/storageservice/ArtifactStoreService.java [hunk 1]
```diff
@@ -114,4 +116,22 @@ default String getUrl(
 
     return uriComponentsBuilder.toUriString();
   }
+
+  default void validateArtifactSizeForTrial(App app, int artifactSize, Long partNumber)
+      throws ModelDBException {
+    if (app.getTrialEnabled()) {
+      if (partNumber != null && partNumber != 0) {
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/artifactStore/storageservice/ArtifactStoreService.java]: we should return multipart upload false in trial mode

**Review by ravishetye — CHANGES_REQUESTED** [review]
this needs update for presigned url flow , I will follow up with you on slack

File: backend/src/main/java/ai/verta/modeldb/experimentRun/ExperimentRunServiceImpl.java [hunk 2]
```diff
@@ -2410,7 +2428,13 @@ public void commitMultipartArtifact(
...
       if (errorMessage != null) {
-        throw new ModelDBException(errorMessage, io.grpc.Status.Code.INVALID_ARGUMENT);
+        throw new ModelDBException(errorMessage, Code.INVALID_ARGUMENT);
+      }
+
+      if (App.getInstance().getTrialEnabled()) {
+        throw new ModelDBException(
+            "Multipart artifact upload not supported on the trial version",
+            Code.FAILED_PRECONDITION);
       }
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/experimentRun/ExperimentRunServiceImpl.java]: Follow the format for errot messages and error code

**Comment by ravishetye** [comment]
When we fetch the temporary credentials let us save them and reuse for this request

File: backend/src/main/java/ai/verta/modeldb/experimentRun/ExperimentRunServiceImpl.java (hunk 2 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/experimentRun/ExperimentRunServiceImpl.java]: it is already resolved on other subsequent PR https://github.com/VertaAI/modeldb/pull/1616

File: backend/src/main/java/ai/verta/modeldb/artifactStore/storageservice/s3/S3Service.java [hunk 3]
```diff
@@ -124,6 +124,52 @@ private Boolean doesObjectExist(String bucketName, String path) throws ModelDBEx
     }
   }
 
+  @Override
+  public GetUrlForArtifact.Response generatePresignedUrlForTrial(
+      String s3Key, String method, long partNumber, String uploadId) throws ModelDBException {
+
+    if (partNumber != 0) {
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/artifactStore/storageservice/s3/S3Service.java]: this should change to 1 based on the way client checks things
https://github.com/VertaAI/modeldb/blob/master/client/verta/verta/_tracking/experimentrun.py#L272

File: backend/src/main/java/ai/verta/modeldb/artifactStore/storageservice/ArtifactStoreService.java (hunk 1 above)
Comment (ravishetye) [reply backend/src/main/java/ai/verta/modeldb/artifactStore/storageservice/ArtifactStoreService.java]: this probably should be 1 https://github.com/VertaAI/modeldb/blob/master/client/verta/verta/_tracking/experimentrun.py#L272

File: backend/src/main/java/ai/verta/modeldb/artifactStore/storageservice/ArtifactStoreService.java [hunk 4]
```diff
@@ -114,4 +120,27 @@ default String getUrl(
...
+                + "Multipart artifact upload not supported on the trial version",
+            Code.RESOURCE_EXHAUSTED);
+      }
+
+      double uploadedArtifactSize = ((double) artifactSize / 1024); // In KB
+      if (app.getMaxArtifactSizeMB() != null
+          && uploadedArtifactSize > ((double) app.getMaxArtifactSizeMB() * 1024)) {
+        throw new ModelDBException(
+            ModelDBConstants.LIMIT_RUN_ARTIFACT_SIZE
+                + "Artifact size more then "
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/artifactStore/storageservice/ArtifactStoreService.java]:
```suggestion
                + "Artifact size more than "
```

File: backend/src/main/java/ai/verta/modeldb/artifactStore/storageservice/ArtifactStoreService.java [hunk 5]
```diff
@@ -114,4 +120,27 @@ default String getUrl(
...
+                + app.getMaxArtifactSizeMB()
+                + " MB not supported",
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/artifactStore/storageservice/ArtifactStoreService.java]:
```suggestion
                + " MB not supported during the trial",
```

File: backend/src/main/java/ai/verta/modeldb/artifactStore/storageservice/ArtifactStoreService.java [hunk 6]
```diff
@@ -114,4 +120,27 @@ default String getUrl(
...
     return uriComponentsBuilder.toUriString();
   }
+
+  default void validateArtifactSizeForTrial(App app, int artifactSize, Long partNumber)
+      throws ModelDBException {
+    if (app.getTrialEnabled()) {
+      if (partNumber != null && partNumber != 0) {
+        throw new ModelDBException(
+            ModelDBConstants.LIMIT_RUN_ARTIFACT_SIZE
+                + "Multipart artifact upload not supported on the trial version",
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/artifactStore/storageservice/ArtifactStoreService.java]:
```suggestion
                + "Multipart artifact upload not supported during the trial",
```

File: backend/src/main/java/ai/verta/modeldb/artifactStore/storageservice/s3/S3Client.java [hunk 7]
```diff
@@ -69,6 +78,19 @@ public S3Client(String cloudBucketName) throws IOException {
...
       initializetWithEnvironment(awsRegion);
     }
+
+    initializeS3SignatureUtil();
+  }
+
+  private void initializeS3SignatureUtil() throws ModelDBException {
+    if (app.getTrialEnabled()) {
+      if (awsCredentials == null) {
+        throw new ModelDBException("awsCredentials should be required for the verta trial");
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/artifactStore/storageservice/s3/S3Client.java]:
```suggestion
        throw new ModelDBException("awsCredentials are required for the verta trial");
```

File: backend/src/main/java/ai/verta/modeldb/artifactStore/storageservice/s3/S3Service.java [hunk 8]
```diff
@@ -124,6 +124,52 @@ private Boolean doesObjectExist(String bucketName, String path) throws ModelDBEx
...
+      throw new ModelDBException(
+          ModelDBConstants.LIMIT_RUN_ARTIFACT_SIZE
+              + "Multipart artifact upload not supported on the trial version",
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/artifactStore/storageservice/s3/S3Service.java]:
```suggestion
              + "Multipart artifact upload not supported during the trial",
```

File: backend/src/main/java/ai/verta/modeldb/artifactStore/storageservice/s3/S3Service.java [hunk 9]
```diff
@@ -124,6 +124,52 @@ private Boolean doesObjectExist(String bucketName, String path) throws ModelDBEx
...
+            .putAllFields(
+                s3Client.getBodyParameterMapForTrialPresignedURL(
+                    s3Key, maxArtifactSize * 1024 * 1024))
+            .build();
+      } else {
+        throw new ModelDBException(
+            ModelDBConstants.LIMIT_RUN_ARTIFACT_SIZE
+                + "Method type ("
+                + method
+                + ") not supported on the trial",
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/artifactStore/storageservice/s3/S3Service.java]:
```suggestion
            ModelDBConstants.LIMIT_RUN_ARTIFACT_SIZE
                + "Method type "
                + method
                + " is not supported during the trial",
```

**Pull request merged**